src/async-wrap.cc

```diff
@@ -126,55 +134,86 @@ RetainedObjectInfo* WrapperInfo(uint16_t class_id, Local<Value> wrapper) {
 // end RetainedAsyncInfo
 
 
-static void EnableHooksJS(const FunctionCallbackInfo<Value>& args) {
+static void SetupHooks(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
-  Local<Function> init_fn = env->async_hooks_init_function();
-  if (init_fn.IsEmpty() || !init_fn->IsFunction())
-    return env->ThrowTypeError("init callback is not assigned to a function");
-  env->async_hooks()->set_enable_callbacks(1);
+
+  if (!args[0]->IsObject())
+    return env->ThrowTypeError("first argument must be an object");
+
+  // All of init, before, after, destroy are supplied by async_hooks
+  // internally, so this should every only be called once. At which time all
+  // the functions should be set. Detect this by checking if init !IsEmpty().
+  CHECK(env->async_hooks_init_function().IsEmpty());
+
+  Local<Object> fn_obj = args[0].As<Object>();
+
+#define GET_HOOK_FN(name)                                                     \
+  Local<Value> name##_v = fn_obj->Get(                                        \
+      env->context(),                                                         \
+      FIXED_ONE_BYTE_STRING(env->isolate(), #name)).ToLocalChecked();         \
+  CHECK(name##_v->IsFunction());                                              \
+  env->set_async_hooks_##name##_function(name##_v.As<Function>());
+  GET_HOOK_FN(init);
```

Comment:
nit: Can you leave a space above this so that it’s more obvious where the macro ends?